Feature/KAS-4348: Resend case to VP

sergiofenoll opened the pull request:
https://kanselarij.atlassian.net/browse/KAS-4348

>[!WARNING]
> This PR isn't ready to be merged yet, there's still a bunch of work to be done

Adds the new submitted piece model and makes some small changes to the SendToVpModal:

- Doctype check Decreet, not Ontwerpdecreet
- Take derived/source files into account when checking what's missing
- Show if both signed and Word file is missing (using suffix `XXX mist (ondertekend en in Word)`

Things that should still be changed here to fulfil the ticket requirements:
- [x] An "opmerkingenveld" needs to be added to the send to VP modal and it should be sent to the backend
- [x] ~~Make sure the button and modal are only shown to admin/secretarie/kabinetsdossierbeheerder of the submitting mandatee~~ (this is now part of https://kanselarij.atlassian.net/browse/KAS-4433)
- [x] The "Send to VP modal" should know what files we're actually sending (on a re-send we don't actually send all the files, the current frontend implementation doesn't know that)

Nice to have things for this feature (but that don't need to be implemented for this ticket specifically):
- [ ] The "Send to VP modal" should probably get the files its sending from the backend, to make sure it accurately reflects what's going to be sent
- [x] When the "Send to VP modal" displays the "Deze bestanden ontbreken nog" message, it should be aware of files it has already previously sent (i.e. if we're just sending a signed PDF but everything else has already been sent, it shouldn't complain about missing all other files)
- [ ] Other stuff?

Related PRs:
- [ ] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/438
- [ ] https://github.com/kanselarij-vlaanderen/vlaams-parlement-sync-service/pull/4

sergiofenoll committed:
chore: check document type Decreet, not Ontwerpdecreet

sergiofenoll committed:
fix: consider derived files in send to VP modal

sergiofenoll committed:
chore: make mimeType "optional" in functions where it's optional

sergiofenoll committed:
chore: only load the document types we need

sergiofenoll committed:
chore: show warning messages when file type is missing

When the PDF is missing: show that the type is missing
When the signed file is missing: show that using (ondertekend)
When the Word is missing: show that using (in Word)
When signed & Word are mssing: show that using (ondertekend en in Word)

sergiofenoll committed:
feat: add SubmittedPiece model

sergiofenoll committed:
chore: explicitly define inverses in parliament models

sergiofenoll committed:
chore: rename publication-remark to remarks

Makes it a more generic translation key that we can reuse elsewhere.

sergiofenoll committed:
feat: add remarks field

sergiofenoll committed:
feat: only show send to VP button if current agendaitem isn't postponed

sergiofenoll committed:
chore: add inverse relation between piece and submitted-piece

sergiofenoll committed:
feat: fetch all pieces and separate pieces that are ready to be sent

All pieces are used to determine what has already been sent, pieces
ready to be sent is what will be sent in the next submission.

sergiofenoll committed:
feat: bust the cache when requesting submitted-piece of piece subclasses

sergiofenoll committed:
chore: sort pieces in send to VP modal

sergiofenoll committed:
chore: show error toast with backend response in send to VP modal

sergiofenoll committed:
chore: disable send to vp modal if there's nothing to send

sergiofenoll committed:
fix: make link between piece and submitted-piece has-many and fix check

The existence of submittedPiece(s) isn't enough to know if a piece
should be sent. We instead need to check the submitted pieces and see if
the files we're ready to send are linked to any of the existing
submitted pieces.

sergiofenoll committed:
feat: move manage-agendaitems check inside agendaitem-controls component

This lets us show the component to all profiles, not only those who can
manage the agendaitems (specifically for sending to VP button)

sergiofenoll committed:
feat: expand canSendToVP check to take into account dossierbeheerders

kobemertens committed:
add components to show status

sergiofenoll committed:
feat: keep track of whether we're sending a complete case to VP

sergiofenoll committed:
chore: add inverse links between parliament models

sergiofenoll committed:
chore: cleanup submission alert a little bit

sergiofenoll committed:
feat: make SendToVPModal and SentToVPModal show files the same way

sergiofenoll committed:
chore: remove stray g from submissions alert component

sergiofenoll committed:
feat: format list of submission activity dates

sergiofenoll committed:
fix: move starting of task

This should hopefully fix cancellation errors being thrown in the CI
pipeline.

tdn commented on app/components/agenda/agendaitem/agendaitem-controls.js:
```diff
@@ -36,12 +36,23 @@ export default class AgendaitemControls extends Component {
-    this.loadCanSendToVP.perform();
+    this.loadDecisionActivity.perform()
+        .then(() => {
+          this.loadCanSendToVP.perform();
```
Do we want to wrap this in the feature flag?

tdn replied:
Nevermind, the function itself checks it.

tdn commented on app/components/agenda/agendaitem/send-to-vp-modal.js:
```diff
@@ -38,20 +51,104 @@ export default class SendToVpModalComponent extends Component {
+        ].join(','),
+        'filter[submission-activity][subcase][:id:]': subcase.id,
+        sort: 'name',
+      })).slice();
```
Why is this `slice()` here?

sergiofenoll replied:
The result of `query{All}` and other Ember Data store related stuff (`await record.thisIsAHasManyRelationship`) is a special kind of Ember array object. `ArrayProxy` or `ManyArray` or `RecordArray`. They may not always support all array-like methods/be iterable. 

When you call `slice()` on it you get a real native JS array. I think in this specific case I was getting errors when iterating over the result on line 99 when no report was present.

tdn commented on app/components/agenda/agendaitem/send-to-vp-modal.js:
```diff
@@ -38,20 +51,104 @@ export default class SendToVpModalComponent extends Component {
+          const signedPieceCopy = await piece.signedPieceCopy;
+          signed = await signedPieceCopy?.file;
+
+          if (submittedPiece.unsignedFile === pdf.uri) {
+            hasSentPdf = true;
+          }
+          if (submittedPiece.wordFile === word?.uri) {
+            hasSentWord = true;
+          }
+          if (submittedPiece.signedFile === signed?.uri) {
+            hasSentSigned = true;
+          }
```
I'm always reluctant to write something like `if (somestatement) { someBoolean = true } else { someBoolean = false }` unless it's inside a loop where you only set the value of `someBoolean` once and then break the loop.

These statements can be rewritten as `someBoolean = someStatement`, so in this case, I'd prefer

```
let hasSentPdf = (submittedPiece.unsignedFile === pdf.uri);
let hasSentWord = (submittedPiece.wordFile === word?.uri);
let hasSentSigned = (submittedPiece.signedFile === signed?.uri);
```

which replaces 12 lines of code by 3.

sergiofenoll replied:
The problem is that we only want to assign in the positive case.

When we submit a piece we only send what hasn't been sent. Take the following case:

- Piece A has PDF and DOCX, we send both to VP
- Piece A gets signed
- Signed version of Piece A gets sent to VP, original PDF and DOCX won't be sent because they haven't changed
- We want to send to VP again and now calculate which pieces have been sent (done in this loop)

In the current implementation, in the first iteration we'll get the submitted piece with PDF and DOCX, mark `hasSentPdf` and `hasSentWord` as true. In the next iteration we'll get the signed piece and mark `hasSentSigned`  as true. Once done, we know all three have been sent and that the piece no longer needs to be sent to the VP.

If we assign like you suggest, in the second iteration we'll find:

```js
let hasSentPdf = (submittedPiece.unsignedFile === pdf.uri); // False, because 2nd submitted piece has no unsignedFile
let hasSentWord = (submittedPiece.wordFile === word?.uri); // False, because 2nd submitted piece has no wordFile
let hasSentSigned = (submittedPiece.signedFile === signed?.uri); // True
```

Which will cause the piece to be erroneously resent.

tdn commented on app/components/agenda/agendaitem/send-to-vp-modal.js:
```diff
@@ -96,25 +215,31 @@ export default class SendToVpModalComponent extends Component {
 
-    const BESLISSINGSFICHE = this.documentTypes.find(
-      (type) => type.uri === CONSTANTS.DOCUMENT_TYPES.BESLISSINGSFICHE
+  async loadDocumentTypes() {
```
This loads the same (or at least an overlapping set of) document types as the `loadData` function. I'd rather see this in a separate task, with the constants then being accessible for the whole component all at once.

tdn commented on app/models/parliament-submission-activity.js:
```diff
@@ -4,7 +4,7 @@ export default class ParliamentSubmissionActivity extends Model {
+  @belongsTo('user', { inverse: null, async: true }) submitter;
 
-  @hasMany('piece', { async: true }) pieces;
+  @hasMany('submitted-piece', { inverse: null, async: true }) submittedPieces;
```
Just a question, since I know this was a point of contention while creating the story. We did land 100% on not persisting the `comment` that goes with the `submission-activity`?

sergiofenoll replied:
No idea, that was said as such in the story
> Wij hoeven (voorlopig toch) dat veld NIET bij te houden in onze data. M.a.w. de inhoud van het opmerkingen veld wordt meegestuurd naar de vlaams-parlement-sync-service bij indiening, maar daarna niet bijgehouden in het datamodel.

and I didn't know who to follow up with about it so I just implemented it as stated.

tdn replied:
so as discussed in standup, I'm adding the comment part to the model anyway, since it was decided while I was away.

tdn reviewed (changes requested):
Some minor code style/clarification comments, but otherwise looks good at first glance.

Pending these changes and a few tests tomorrow, I think this is close to being mergeable.

sergiofenoll committed:
chore: extract loading of document types into single method

sergiofenoll committed:
chore: remove send to VP permission from secretarie/KB

sergiofenoll committed:
docs: document strange behaviour in format-list component

sergiofenoll committed:
fix: remove unused query in route

sergiofenoll committed:
fix: don't try to do VP checks if there's no subcase

This is the case in the "approval for meeting" agendaitem, which every
agenda has and otherwise causes errors.

sergiofenoll committed:
fix: don't try to do VP checks if there is no subcase

sergiofenoll committed:
chore: remove trim in FormatList

The literal parts also include the needed whitespace, we should keep that.

sergiofenoll committed:
chore: sort parliament submission activities

tdn committed:
Merge pull request #1973 from kanselarij-vlaanderen/feature/KAS-4433-send-to-vp-role-permissions

Feature/kas 4433 send to vp role permissions

tdn committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-4348-resend-case-to-vp

tdn committed:
added comment to model

tdn committed:
Merge branch 'feature/KAS-4348-resend-case-to-vp' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-4350-weergave-submission-activities

tdn committed:
Merge pull request #1974 from kanselarij-vlaanderen/feature/KAS-4350-weergave-submission-activities

feature/KAS 4350 weergave submission activities

tdn committed:
added comment to sent-to-vp modal

tdn committed:
corrected a copy-paste status

tdn committed:
Made sure isBeingHandledByFP is displayed correctly
